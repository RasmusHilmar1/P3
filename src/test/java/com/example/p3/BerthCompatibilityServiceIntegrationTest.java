package com.example.p3;

import com.example.p3.model.Berth;
import com.example.p3.model.Boat;
import com.example.p3.model.BerthWithCompatibility;
import com.example.p3.service.BerthCompatibilityService;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;


import java.util.List;
import java.util.stream.Collectors;

@SpringBootTest
public class BerthCompatibilityServiceIntegrationTest {

    @Autowired
    private BerthCompatibilityService berthCompatibilityService;

    private Boat boat;

    @Test
    public void testCompatibilityReport() {
        // Set up a boat with specific dimensions for testing
        Boat boat = new Boat();
        boat.setLength(13.0); // Length of the boat
        boat.setWidth(3.5);  // Width of the boat

        // Call the method to find compatible berths from the real database
        List<BerthWithCompatibility> compatibleBerths = berthCompatibilityService.findCompatibleBerthsWithScore(boat);

        // Generate the compatibility report and verify using ApprovalTests
        String report = compatibleBerths.stream()
                .map(berthWithCompatibility -> {
                    Berth berth = berthWithCompatibility.getBerth();
                    double compatibilityScore = berthWithCompatibility.getCompatibilityScore();
                    String compatibilitySquare = getCompatibilitySquare(compatibilityScore);
                    return String.format("Berth: %s, Length: %.2f, Width: %.2f, Availability: %d, " +
                                    "Compatibility Score: %.2f, %s",
                            berth.getName(), berth.getLength(), berth.getWidth(),
                            berth.getAvailability(), compatibilityScore, compatibilitySquare);
                })
                .collect(Collectors.joining("\n"));

        // Use ApprovalTests to verify the report
        Approvals.verify(report);
    }


    // Method to get the colored square emoji based on compatibility score
    private static String getCompatibilitySquare(double compatibility) {
        if (compatibility >= 90) {
            return "ðŸŸ©"; // Green square for high compatibility
        } else if (compatibility >= 70) {
            return "ðŸŸ¨"; // Yellow square for medium compatibility
        } else if (compatibility >= 50) {
            return "ðŸŸ§"; // Orange square for low compatibility
        } else if (compatibility > 0) {
            return "ðŸŸ¥"; // Red square for very low compatibility
        } else {
            return "ðŸŸª"; // Purple square for incompatible
        }
    }
}
