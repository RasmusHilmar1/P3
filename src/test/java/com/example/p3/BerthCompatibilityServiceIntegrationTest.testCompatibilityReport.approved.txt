Berth: VN08, Length: 14,10, Width: 4,00, Availability: 1, Compatibility Score: 97,01, ğŸŸ©
Berth: NB14, Length: 15,00, Width: 4,00, Availability: 1, Compatibility Score: 93,80, ğŸŸ©
Berth: VN03, Length: 14,20, Width: 4,80, Availability: 1, Compatibility Score: 86,13, ğŸŸ¨
Berth: VN04, Length: 14,20, Width: 4,80, Availability: 1, Compatibility Score: 86,13, ğŸŸ¨
Berth: VN13, Length: 14,00, Width: 5,20, Availability: 1, Compatibility Score: 81,58, ğŸŸ¨
Berth: VN11, Length: 14,10, Width: 5,20, Availability: 1, Compatibility Score: 81,22, ğŸŸ¨
Berth: NM36, Length: 14,50, Width: 5,10, Availability: 1, Compatibility Score: 81,11, ğŸŸ¨
Berth: NM37, Length: 14,50, Width: 5,10, Availability: 1, Compatibility Score: 81,11, ğŸŸ¨
Berth: NM38, Length: 14,50, Width: 5,10, Availability: 1, Compatibility Score: 81,11, ğŸŸ¨
Berth: NB61, Length: 15,00, Width: 5,80, Availability: 1, Compatibility Score: 70,11, ğŸŸ¨